dataset/shape.cpp

```diff
@@ -141,4 +141,232 @@ Dataset resize(const DatasetConstView &d, const Dim dim,
+    if (copy.dims().contains(labels[l])) {
+      reshaped = Variable(copy.slice({labels[l], 0}));
+      for (int32_t i = 1; i < from_dims.shape()[l]; ++i)
+        reshaped = join_edges(VariableView(reshaped),
+                              copy.slice({labels[l], i}), labels.back());
+      // Update copy
+      copy = reshaped;
+    }
+  }
+  copy.rename(labels.back(), to_dim);
+  return copy;
+}
+
+/// Check if at keast one of the variable's dimensions is bin edges
+bool has_bin_edges(const VariableConstView &var, const Dimensions &array_dims) {
+  for (const auto dim : var.dims().labels())
+    if (is_bin_edges(var, array_dims, dim))
+      return true;
+  return false;
+}
+
+} // end anonymous namespace
+
+/// Reshape a single dimension into multiple dimensions:
+/// ['x': 6] -> ['y': 2, 'z': 3]
+DataArray reshape(const DataArrayConstView &a, const Dim from_dim,
+                  const Dimensions &to_dims) {
+  // Make sure that new dims do not already exist in data dimensions, apart
+  // apart from the old dim (i.e. old dim can be re-used)
+  auto old_dims = a.dims();
+  for (const auto dim : to_dims.labels())
+    if (old_dims.contains(dim) && dim != from_dim)
+      throw except::DimensionError(
+          "Reshape: new dimensions cannot contain labels that "
+          "already exist in the DataArray.");
+
+  auto reshaped =
+      DataArray(reshape(a.data(), stack_dims(old_dims, from_dim, to_dims)));
+
+  for (auto &&[name, coord] : a.coords()) {
+    if (is_bin_edges(coord, old_dims, from_dim))
+      reshaped.coords().set(name, slice_and_stack(coord, from_dim, to_dims));
+    else
+      reshaped.coords().set(
+          name, reshape(coord, stack_dims(coord.dims(), from_dim, to_dims)));
+  }
+
+  for (auto &&[name, attr] : a.attrs())
+    if (is_bin_edges(attr, old_dims, from_dim))
+      reshaped.attrs().set(name, slice_and_stack(attr, from_dim, to_dims));
+    else
+      reshaped.attrs().set(
+          name, reshape(attr, stack_dims(attr.dims(), from_dim, to_dims)));
+
+  // Note that we assume bin-edge masks do not exist
+  for (auto &&[name, mask] : a.masks())
+    reshaped.masks().set(
+        name, reshape(mask, stack_dims(mask.dims(), from_dim, to_dims)));
+
+  return reshaped;
+}
+
+/// Reshape multiple dimensions into a single dimension:
+/// ['y', 'z'] -> ['x']
+DataArray reshape(const DataArrayConstView &a,
+                  const std::vector<Dim> &from_dims, const Dim to_dim) {
+  auto old_dims = a.dims();
+  for (const auto dim : from_dims)
+    if (!old_dims.contains(dim))
+      throw except::DimensionError(
+          "Reshape: dimension to be reshaped not found in DataArray.");
+
+  Dimensions stacked_dims;
+  for (const auto dim : from_dims)
+    stacked_dims.addInner(dim, old_dims[dim]);
+
+  auto reshaped = DataArray(
+      reshape(a.data(), unstack_dims(old_dims, stacked_dims, to_dim)));
+
+  for (auto &&[name, coord] : a.coords()) {
+    if (has_bin_edges(coord, old_dims)) {
```

Comment:
Is this the correct check? Shouldn't it only check if one of the `from_dims` is a bin-edge dim of `coord`?